### ISSUE TITLE
Error: Unable to start the storage emulator v5.10.0

### ISSUE THREAD

**Issue by spottedmahn** [Azure/azure-storage-net#933]
### Which service(blob, file, queue, table) does this issue concern?
Storage Emulator

### Which version of the SDK was used?
No clue.  I see MSFT Azure Storage Emulator - v5.10 in add/remove programs.

### What problem was encountered?
Trying to follow this guide, [Exercise - Create and test a simple Azure Function locally with Visual Studio](https://docs.microsoft.com/en-us/learn/modules/develop-test-deploy-azure-functions-with-visual-studio/3-exercise-develop-and-test-azure-functions-locally), in VS 2019 and I'm getting when I start my app up:

![image](https://user-images.githubusercontent.com/14079228/65350576-05d5ba80-dbb4-11e9-8626-7c6eb71a7cdf.png)

I've **tried** starting in a **admin cmd prompt** but it is still failing (with no details as to why)

>C:\Program Files (x86)\Microsoft SDKs\Azure\Storage Emulator>AzureStorageEmulator.exe start
Windows Azure Storage Emulator 5.10.0.0 command line tool
Error: Unable to start the storage emulator.

I see it running the sys tray but I can't start it there either.

![image](https://user-images.githubusercontent.com/14079228/65350674-3ddcfd80-dbb4-11e9-94c8-1526a48fc868.png)

**Comment by spottedmahn** [comment]
>C:\Program Files (x86)\Microsoft SDKs\Azure\Storage Emulator>AzureStorageEmulator.exe init
Windows Azure Storage Emulator 5.10.0.0 command line tool
Found SQL Instance (localdb)\MSSQLLocalDB.
Creating database AzureStorageEmulatorDb510 on SQL instance '(localdb)\MSSQLLocalDB'.

>Granting database access to user CEI.DOMAIN\mdepouw.
Database access for user CEI.DOMAIN\mdepouw was granted.

>Initialization successful. The storage emulator is now ready for use.
The storage emulator was successfully initialized and is ready to use.

**Comment by spottedmahn** [comment]
I continued to run commands and try stuff.  Now when I run `init` I get:

>C:\Program Files (x86)\Microsoft SDKs\Azure\Storage Emulator>AzureStorageEmulator.exe init
Windows Azure Storage Emulator 5.10.0.0 command line tool
Found SQL Instance (localdb)\MSSQLLocalDB.
Creating database AzureStorageEmulatorDb510 on SQL instance '(localdb)\MSSQLLocalDB'.
Cannot create database 'AzureStorageEmulatorDb510' : The database 'AzureStorageEmulatorDb510' does not exist. Supply a valid database name. To see available databases, use sys.databases..
One or more initialization actions have failed. Resolve these errors before attempting to run the storage emulator again.
Error: Cannot create database 'AzureStorageEmulatorDb510' : The database 'AzureStorageEmulatorDb510' does not exist. Supply a valid database name. To see available databases, use sys.databases..

**Comment by spottedmahn** [comment]
I manually created the DB but it doesn't seem to create the schema when running `init`

>C:\Program Files (x86)\Microsoft SDKs\Azure\Storage Emulator>AzureStorageEmulator.exe init
Windows Azure Storage Emulator 5.10.0.0 command line tool
Found SQL Instance (localdb)\MSSQLLocalDB.
Creating database AzureStorageEmulatorDb510 on SQL instance '(localdb)\MSSQLLocalDB'.

>Granting database access to user CEI.DOMAIN\mdepouw.
>Database access for user CEI.DOMAIN\mdepouw was granted.

Initialization successful. The storage emulator is now ready for use.
The storage emulator was successfully initialized and is ready to use.

![image](https://user-images.githubusercontent.com/14079228/65352474-c2ca1600-dbb8-11e9-8ca2-9b8f91c7a41e.png)

**Comment by abatishchev** [comment]
Same error here:

```
>sqllocaldb i
MSSQLLocalDB

>AzureStorageEmulator.exe init
Windows Azure Storage Emulator 5.10.0.0 command line tool
Found SQL Instance (localdb)\MSSQLLocalDB.
Creating database AzureStorageEmulatorDb510 on SQL instance '(localdb)\MSSQLLocalDB'.
Cannot create database 'AzureStorageEmulatorDb510' : The database 'AzureStorageEmulatorDb510' does not exist. Supply a valid database name. To see available databases, use sys.databases..
One or more initialization actions have failed. Resolve these errors before attempting to run the storage emulator again.
Error: Cannot create database 'AzureStorageEmulatorDb510' : The database 'AzureStorageEmulatorDb510' does not exist. Supply a valid database name. To see available databases, use sys.databases..
```

**Comment by JustinGrote** [comment]
Are you just using the RTM download? You have to upgrade to CU17, I think around CU5 there's a bug where it tries to initialize into the empty users directly.

**Comment by spottedmahn** [comment]
13.0.4001 which is apparently [2016 SP1](https://support.microsoft.com/en-us/help/3177312/sql-server-2016-build-versions).  I don't manage that, isn't VS responsible for that? 🤔

![image](https://user-images.githubusercontent.com/14079228/76809836-7ad02880-67c2-11ea-9e32-591b479542af.png)

**Comment by abatishchev** [comment]
You need to install CU yourself. Installing the latest fixed the problem for me.

**Comment by JustinGrote** [comment]
Yes, it's a massive pain. I've since switched to Azurite now that the storage emulator is deprecated in favor of it and it's a much better experience. If you need table support download the cosmosDB emulator.

**Comment by 1TheMuffinMan** [comment]
I'm having the same issue all of sudden. Not getting anywhere.

**Comment by lpgoncalves** [comment]
Same error, tried as admin, init first, close apps that would use same doors, restart windows, but still doesn't work.

**Comment by 1TheMuffinMan** [comment]
One thing to try, is download the latest Azure storage tools and then ensure your project is using them. The value you use to choose the tooling is esoteric and seemingly not documented anywhere.



1. Navigate to `%appdata%`, local\AzureFunctionsTools.
2. Open feed-v3.json in a text editor
3. Note the keys `v3`, `v2.preview` etc
4. Find whatever corrolates to the newest version and make sure that string is placed in the  `<AzureFunctionsVersion>v3</AzureFunctionsVersion>` in the `csproj` file.

![image](https://user-images.githubusercontent.com/4889283/88096988-5effd900-cb4c-11ea-9fce-f59be3a689c2.png)

![image](https://user-images.githubusercontent.com/4889283/88097016-6921d780-cb4c-11ea-839f-96603fad3424.png)

![image](https://user-images.githubusercontent.com/4889283/88097056-750d9980-cb4c-11ea-9892-e50cc5a798e0.png)

So in summary, updating the latest azure storage emulator won't automatically use it in your project.

**Comment by ishepherd** [comment]
My fix was:
1. Run `sqllocaldb s` to start the MSSQLLocalDB database
2. `sqllocaldb i MSSQLLocalDB` showed it was running an old version, `14.0.1000.169`
3. [Install the latest Cumulative Update for SQL Server](https://support.microsoft.com/en-us/topic/kb5000685-cumulative-update-23-for-sql-server-2017-22b653c7-8487-4564-9db2-b5c1bd465145)
4. `sqllocaldb i MSSQLLocalDB` now showed `14.0.3381.3` (I also used the sqllocaldb subcommands to restart it manually)

At this point `AzureStorageEmulator.exe init` succeeded (I stopped getting "Cannot create database 'AzureStorageEmulatorDb510'")

----

`AzureStorageEmulator.exe start` still fails. Now produces

```
Windows Azure Storage Emulator 5.10.0.0 command line tool

Unhandled Exception: System.TimeoutException: Unable to open wait handle.
   at Microsoft.WindowsAzure.Storage.Emulator.Controller.EmulatorProcessController.InternalWaitForStorageEmulator(Int32 timeoutInMilliseconds)
   at Microsoft.WindowsAzure.Storage.Emulator.Controller.EmulatorProcessController.EnsureRunning(Int32 timeoutInMilliseconds)
   at Microsoft.WindowsAzure.Storage.Emulator.Commands.StartCommand.RunCommand()
   at Microsoft.WindowsAzure.Storage.Emulator.Program.Main(String[] args)
```

`AzureStorageEmulator.exe status` suggests it wants to use ports 10000-10002 and `netstat -abno` suggests Docker is already using 10000, 10001

I will try moving the emulator as per the answer @lpgoncalves found: https://stackoverflow.com/a/59854137

**Comment by osse-o** [comment]
This worked for me:
`netsh http add iplisten 127.0.0.1`
found here: [https://onlyhuman.dk/troubleshooting/2019/09/27/Azure_Storage_Emulator_Network_location_cannot_be_reached/](url)

**Comment by karvantestandtracenhsuk** [comment]
I uninstalled my IIS and that starts working again. ( I had the emulator working till I installed the IIS and it stopped working suddenly).

**Comment by KarvanJafi** [comment]
I had it working fine till I installed IIS on windows 10 and my storage emulator stopped working. I uninstalled IIS and it's working fine now.

**Comment by JustinGrote** [comment]
@KarvanJafi have you tried Azurite? It has table support now and there's little reason to be using the "heavy" old storage emulator anymore.

**Comment by asankagarusinghe** [comment]
@spottedmahn 

sometimes, we have to recover init processes involving in the 'AzureStorageEmulatorDb*' database. 
Always trying to follow this approach. worked for me in most occasions 

Delete the SQL instance:

```
sqllocaldb stop MSSQLLocalDB
sqllocaldb delete MSSQLLocalDB
```
Now delete the following database files:

%USERPROFILE%\AzureStorageEmulatorDb*.mdf
%USERPROFILE%\AzureStorageEmulatorDb*.ldf
Finally, recreate the sql instance:

sqllocaldb start MSSQLLocalDB

explained here
https://stackoverflow.com/questions/29239053/azure-storage-emulator-fails-after-adding-vs-2015-sdk

**Comment by pkaushik23** [comment]
> My fix was:
> 
> 1. Run `sqllocaldb s` to start the MSSQLLocalDB database
> 2. `sqllocaldb i MSSQLLocalDB` showed it was running an old version, `14.0.1000.169`
> 3. [Install the latest Cumulative Update for SQL Server 14 ('2017')](https://support.microsoft.com/en-us/topic/kb5000685-cumulative-update-23-for-sql-server-2017-22b653c7-8487-4564-9db2-b5c1bd465145)
> 4. `sqllocaldb i MSSQLLocalDB` now showed `14.0.3381.3` (I also used the sqllocaldb subcommands to restart it manually)
> 
> At this point `AzureStorageEmulator.exe init` succeeded (I stopped getting "Cannot create database 'AzureStorageEmulatorDb510'")
> 
> `AzureStorageEmulator.exe start` still fails. Now produces
> 
> ```
> Windows Azure Storage Emulator 5.10.0.0 command line tool
> 
> Unhandled Exception: System.TimeoutException: Unable to open wait handle.
>    at Microsoft.WindowsAzure.Storage.Emulator.Controller.EmulatorProcessController.InternalWaitForStorageEmulator(Int32 timeoutInMilliseconds)
>    at Microsoft.WindowsAzure.Storage.Emulator.Controller.EmulatorProcessController.EnsureRunning(Int32 timeoutInMilliseconds)
>    at Microsoft.WindowsAzure.Storage.Emulator.Commands.StartCommand.RunCommand()
>    at Microsoft.WindowsAzure.Storage.Emulator.Program.Main(String[] args)
> ```
> 
> `AzureStorageEmulator.exe status` suggests it wants to use ports 10000-10002 and `netstat -abno` suggests Docker is already using 10000, 10001
> 
> I will try moving the emulator as per the answer @lpgoncalves found: https://stackoverflow.com/a/59854137

changing port worked for me. Thanks.

**Comment by borregs** [comment]
Troubleshooting this was so exhausting more so since I think it has many root causes and not necessarily a set solution.

My solution (or problem more so) was checking all the listening services and removed or configured the ones that where using localhost:10000-10002 ports. I do want to add that I first tried modifying the ` AzureStorageEmulator.exe.config ` and tried to have it use different ports but this caused some of my applications in development to break. So I wouldn't suggest that approach although it seems to have worked for some.

The service that was messing with blob's ports was `Telerik Mocking`

**Comment by ToMakeSense** [comment]
Same there, I am using win10 21H1, the following possible situations have been ruled out:
1. The localdb is fine, I could connect it via `sqllocaldb` utility and SSMS. 
2. Removing/manually creating the `AzureStorageEmulator510.mdf` also didn't work.
3. the ports1000-10002 were not in use by other App.

see the output in the console
```cmd
C:\Program Files (x86)\Microsoft SDKs\Azure\Storage Emulator>AzureStorageEmulator.exe init
Windows Azure Storage Emulator 5.10.0.0 command line tool
Empty SQL Instance. Autodetecting SQL Instance to use.
Looking for a LocalDB Installation.
Probing SQL Instance: '(localdb)\MSSQLLocalDB'.
Caught exception while probing for SQL endpoint. 在与 SQL Server 建立连接时出现与网络相关的或特定于实例的错误。未找到或无法访问服务器。请验证实例名称是否正确并且 SQL Server 已配置为允许远程连接。 (provider: SQL Network Interfaces, error: 25 - 连接字符串无效)
Number of SqlErrors Reported: 1
SqlError: System.Data.SqlClient.SqlError: 在与 SQL Server 建立连接时出现与网络相关的或特定于实例的错误。未找到或无法访问服务器。请验证实例名称是否正确并且 SQL Server 已配置为允许远程连接。 (provider: SQL Network Interfaces, error: 25 - 连接字符串无效)
Could not find a LocalDB Installation.
Probing SQL Instance: 'localhost\SQLExpress'.
Caught exception while probing for SQL endpoint. 在与 SQL Server 建立连接时出现与网络相关的或特定于实例的错误。未找到或无法访问服务器。请验证实例名称是否正确并且 SQL Server 已配置为允许远程连接。 (provider: SQL Network Interfaces, error: 26 - 定位指定的服务器/实例时出错)
Number of SqlErrors Reported: 1
SqlError: System.Data.SqlClient.SqlError: 在与 SQL Server 建立连接时出现与网络相关的或特定于实例的错误。未找到或无法访问服务器。请验证实例名称是否正确并且 SQL Server 已配置为允许远程连接。 (provider: SQL Network Interfaces, error: 26 - 定位指定的服务器/实例时出错)
No available SQL Instance was found.
One or more initialization actions have failed. Resolve these errors before attempting to run the storage emulator again.
Error: No available SQL Instance was found.

C:\Program Files (x86)\Microsoft SDKs\Azure\Storage Emulator>sqllocaldb v
Microsoft SQL Server 2016 (13.1.4001.0)
Microsoft SQL Server 2016 (13.1.4001.0)

C:\Program Files (x86)\Microsoft SDKs\Azure\Storage Emulator>netstat -ano | findstr 10000

C:\Program Files (x86)\Microsoft SDKs\Azure\Storage Emulator>netstat -ano | findstr 10001

C:\Program Files (x86)\Microsoft SDKs\Azure\Storage Emulator>netstat -ano | findstr 10002

```
And also the error log in `C:\Users\XXXX\AppData\Local\AzureStorageEmulator\Logs`
```log
2022/1/9 22:26:17 [Error] [ActivityId=00000000-0000-0000-0000-000000000000] Configuration Parameter: Nephos.MaskClientIPAddressesInLogs not found
2022/1/9 22:26:17 [Error] [ActivityId=00000000-0000-0000-0000-000000000000] Hit Exception while reading Nephos.MaskClientIPAddressesInLogs setting from ACS: System.Exception: Configuration Parameter: Nephos.MaskClientIPAddressesInLogs not found ---> System.NullReferenceException: 未将对象引用设置到对象的实例。
   在 Microsoft.WindowsAzure.Storage.Emulator.HttpServiceHost.GetConfigurationParameter(String name)
   --- 内部异常堆栈跟踪的结尾 ---
   在 Microsoft.WindowsAzure.Storage.Emulator.HttpServiceHost.GetConfigurationParameter(String name)
   在 Microsoft.Cis.Services.Nephos.Common.Protocols.Rest.ConfigurationHelper.GetConfigSetting[T](IServiceEntrySink serviceEntrySink, String configParam, String configDescription)
   在 Microsoft.Cis.Services.Nephos.Common.Protocols.Rest.ConfigurationHelper.GetIPAddressMaskSetting(IServiceEntrySink sink)
2022/1/9 22:26:17 [Error] [ActivityId=00000000-0000-0000-0000-000000000000] Configuration Parameter: Nephos.MaskClientIPAddressesInLogs not found
2022/1/9 22:26:17 [Error] [ActivityId=00000000-0000-0000-0000-000000000000] Hit Exception while reading Nephos.MaskClientIPAddressesInLogs setting from ACS: System.Exception: Configuration Parameter: Nephos.MaskClientIPAddressesInLogs not found ---> System.NullReferenceException: 未将对象引用设置到对象的实例。
   在 Microsoft.WindowsAzure.Storage.Emulator.HttpServiceHost.GetConfigurationParameter(String name)
   --- 内部异常堆栈跟踪的结尾 ---
   在 Microsoft.WindowsAzure.Storage.Emulator.HttpServiceHost.GetConfigurationParameter(String name)
   在 Microsoft.Cis.Services.Nephos.Common.Protocols.Rest.ConfigurationHelper.GetConfigSetting[T](IServiceEntrySink serviceEntrySink, String configParam, String configDescription)
   在 Microsoft.Cis.Services.Nephos.Common.Protocols.Rest.ConfigurationHelper.GetIPAddressMaskSetting(IServiceEntrySink sink)
2022/1/9 22:26:17 [Error] [ActivityId=00000000-0000-0000-0000-000000000000] Configuration Parameter: Nephos.MaskClientIPAddressesInLogs not found
2022/1/9 22:26:17 [Error] [ActivityId=00000000-0000-0000-0000-000000000000] Hit Exception while reading Nephos.MaskClientIPAddressesInLogs setting from ACS: System.Exception: Configuration Parameter: Nephos.MaskClientIPAddressesInLogs not found ---> System.NullReferenceException: 未将对象引用设置到对象的实例。
   在 Microsoft.WindowsAzure.Storage.Emulator.HttpServiceHost.GetConfigurationParameter(String name)
   --- 内部异常堆栈跟踪的结尾 ---
   在 Microsoft.WindowsAzure.Storage.Emulator.HttpServiceHost.GetConfigurationParameter(String name)
   在 Microsoft.Cis.Services.Nephos.Common.Protocols.Rest.ConfigurationHelper.GetConfigSetting[T](IServiceEntrySink serviceEntrySink, String configParam, String configDescription)
   在 Microsoft.Cis.Services.Nephos.Common.Protocols.Rest.ConfigurationHelper.GetIPAddressMaskSetting(IServiceEntrySink sink)
```

**Comment by mrp100461** [comment]
I found that Node.js was listening on port 100000

**Comment by spottedmahn** [comment]
closing as stale / no activity
